x/gamm/keeper/msg_server_test.go

```diff
@@ -356,3 +361,158 @@ func (suite *KeeperTestSuite) TestExitPool_Events() {
+		},
+		{
+			name: "error: attempt to migrate shares from non-existent pool",
+			param: param{
+				sender:                defaultAccount,
+				sharesToMigrateDenom:  "gamm/pool/1000",
+				sharesToMigrateAmount: defaultGammShares.Amount,
+				poolIdEntering:        2,
+			},
+			sharesToCreate: defaultGammShares.Amount,
+			expectedErr:    fmt.Errorf("pool with ID %d does not exist", 1000),
+		},
+		{
+			name: "error: attempt to migrate more shares than the user has",
+			param: param{
+				sender:                defaultAccount,
+				sharesToMigrateDenom:  defaultGammShares.Denom,
+				sharesToMigrateAmount: defaultGammShares.Amount.Add(sdk.NewInt(1)),
+				poolIdEntering:        2,
+			},
+			sharesToCreate: defaultGammShares.Amount,
+			expectedErr:    sdkerrors.Wrap(sdkerrors.ErrInsufficientFunds, fmt.Sprintf("%s is smaller than %s", defaultGammShares, invalidGammShares)),
+		},
+	}
+
+	for _, test := range tests {
+		suite.SetupTest()
+		msgServer := keeper.NewBalancerMsgServerImpl(suite.App.GAMMKeeper)
+
+		// Prepare both balancer and concentrated pools
+		suite.FundAcc(test.param.sender, defaultAccountFunds)
+		gammPoolId := suite.PrepareBalancerPoolWithCoins(sdk.NewCoin("eth", sdk.NewInt(100000000000)), sdk.NewCoin("usdc", sdk.NewInt(100000000000)))
+		gammPool, err := suite.App.GAMMKeeper.GetPoolAndPoke(suite.Ctx, gammPoolId)
+		suite.Require().NoError(err)
+		clPool := suite.PrepareConcentratedPool()
+
+		// Note gamm and cl pool addresses
+		gammPoolAddress := gammPool.GetAddress()
+		clPoolAddress := clPool.GetAddress()
+
+		// Join gamm pool to create gamm shares directed in the test case
+		_, _, err = suite.App.GAMMKeeper.JoinPoolNoSwap(suite.Ctx, test.param.sender, gammPoolId, test.sharesToCreate, sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(999999999999999)), sdk.NewCoin("usdc", sdk.NewInt(999999999999999))))
+		suite.Require().NoError(err)
+
+		// Note gamm pool balance after joining gamm pool
+		gammPoolEthBalancePostJoin := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, ETH)
+		gammPoolUsdcBalancePostJoin := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, USDC)
+
+		// Create migrate message
+		sharesToMigrate := sdk.NewCoin(test.param.sharesToMigrateDenom, test.param.sharesToMigrateAmount)
+		msg := &balancer.MsgMigrateSharesToFullRangeConcentratedPosition{
+			Sender:          test.param.sender.String(),
+			SharesToMigrate: sharesToMigrate,
+			PoolIdEntering:  test.param.poolIdEntering,
+		}
+
+		// Reset event counts to 0 by creating a new manager.
+		suite.Ctx = suite.Ctx.WithEventManager(sdk.NewEventManager())
+		suite.Require().Equal(0, len(suite.Ctx.EventManager().Events()))
+
+		// Migrate the user's gamm shares to a full range concentrated liquidity position
+		resp, err := msgServer.MigrateSharesToFullRangeConcentratedPosition(sdk.WrapSDKContext(suite.Ctx), msg)
+		if test.expectedErr != nil {
+			suite.Require().Error(err)
+			suite.Require().ErrorContains(err, test.expectedErr.Error())
```

Comment:
Consider validating here that the gamm pool still has shares while CL pool does not have a position created

Reply:
https://github.com/osmosis-labs/osmosis/pull/4051/commits/79009330eca4e9c173a63ee752de127812578a2e